app/models/reservation.rb

```diff
@@ -17,17 +17,23 @@ class Reservation < ApplicationRecord # :nodoc:
   validates :checkin_at, presence: true
   validates :checkout_at, presence: true
   validates :book_code, presence: true, uniqueness: true
-
-  # Callbacks
-  before_create :assigned_room, if: -> { room.unavailable? }
+  validate :room_must_be_available
 
   private
 
-  def assigned_room
+  def room_must_be_available # rubocop:disable Metrics/AbcSize
+    return if room.blank? || checkin_at.blank? || checkout_at.blank?
```

Comment:
Maybe you could move this condition to the `validate` method using the `if` conditional.